http.go

```diff
@@ -1148,6 +1149,24 @@ func (req *Request) Write(w *bufio.Writer) error {
 		}
 		req.Header.SetHostBytes(host)
 		req.Header.SetRequestURIBytes(uri.RequestURI())
+
+		if len(uri.username) > 0 {
+			// RequestHeader.SetBytesKV only uses RequestHeader.bufKV.key
+			// So we are free to use RequestHeader.bufKV.value as a scratch pad for
+			// the base64 encoding.
+			nl := len(uri.username) + len(uri.password) + 1
+			tl := nl + base64.StdEncoding.EncodedLen(nl)
+			if tl > cap(req.Header.bufKV.value) {
+				req.Header.bufKV.value = make([]byte, 0, tl)
+			}
+			buf := req.Header.bufKV.value[:0]
+			buf = append(buf, uri.username...)
+			buf = append(buf, strColon...)
+			buf = append(buf, uri.password...)
+			buf = buf[:tl]
+			base64.StdEncoding.Encode(buf[nl:], buf[:nl])
+			req.Header.SetBytesKV(strAuthorization, buf[nl:])
```

Comment:
I agree this is all a bit dirty to use the scratch buffer of the `req.Header`. But the alternative would be to have another `sync.Pool` for `[]byte` just for this.

Reply:
https://go-review.googlesource.com/c/go/+/37639/1/src/encoding/base64/base64.go

Reply:
That would change these lines:
```go
buf = buf[:tl]
base64.StdEncoding.Encode(buf[nl:], buf[:nl])
```
into:
```go
buf = base64.StdEncoding.AppendEncode(buf, buf[:nl])
```
but it wouldn't help avoiding with the allocation.